### ISSUE TITLE
Contract test fails in macos without jit

### ISSUE THREAD

**Issue by pmatos** [racket/racket#2964]
A contract test is failing without jit on macos:
https://github.com/LinkiTools/racket/runs/346077520#step:10:43

```
FAILED recursive-contract-not-too-slow
       got #f
  expected #t
FAILED recursive-contract-not-too-slow rewrite-to-add-opt/c
       got #f
  expected #t
```

I haven't looked at it but the `slow` in the name might mean this is dependent on the cpu speed. I will take a look at it but I am starting to think we should not have timing in tests and if we do use timing, we should attempt to use cpu time, instead of elapsed time. Again, I haven't looked at this test yet, so it might actually not be time related.

**Comment by samth** [comment]
The name of that test suggests that it's just about not having the JIT is slower.

**Comment by samth** [comment]
One possibility is to increase the timeout when the `CI` environment variable is set. But maybe @rfindler has other thoughts?

**Comment by rfindler** [comment]
This test uses CPU time already, not real time. It is testing for an algorithmic running time bug. I welcome suggestions.

**Comment by mflatt** [comment]
For the core Racket tests, I recently introduced a `PLT_RUN_UNRELIABLE_TESTS` environment variable to guard tests that I don't know how to make reliable, but that I'd like to keep checking when I run tests manually. The environment variable's value is comma-separated list, and currently there are checks for `timing` (in the sense of race conditions, but maybe it makes sense for expecting a certain level of runtime performance) and `locale`. I have `setenv PLT_RUN_UNRELIABLE_TESTS timing,locale` in my `.cshrc`.

I can't think of a way to make this test reliable, so maybe treat it the same way and enable it only if `PLT_RUN_UNRELIABLE_TESTS` includes `timing` or some new tag?

**Comment by samth** [comment]
@rfindler would 10 seconds also be enough to check that the algorithmic time bug is avoided? The issue is that everything is much slower without the JIT and so it takes more than 5 seconds.

**Comment by mflatt** [comment]
Another possibility: If it's a question of asymptotic running time (which I didn't get before), the core tests have successfully used the strategy of comparing run times for N, 2N, 4N, etc. up to some reasonable size.

**Comment by rfindler** [comment]
That sounds like a good idea!

**Comment by pmatos** [comment]
Unfortunately this is still failing:
https://github.com/LinkiTools/racket/runs/348950642#step:10:43

Rebased over https://github.com/racket/racket/commit/c8877055da6ff44fa562f0708f49e14b83c9c823

**Comment by mflatt** [comment]
@pmatos No idea... Can you try https://github.com/mflatt/racket/commit/4be30769cbebc48fa7ca81301972013c2cdc94aa ? Maybe the information it reports on relative run time will help us see the problem.

**Comment by pmatos** [comment]
@mflatt Rebasing over your commit it fails only on 3m/macos:
https://github.com/LinkiTools/racket/runs/355970552#step:10:43

```
FAILED recursive-contract-not-too-slow
       got (632 5698)
  expected #f
FAILED recursive-contract-not-too-slow rewrite-to-add-opt/c
       got (618 5703)
```

**Comment by pmatos** [comment]
I just noticed (https://github.com/LinkiTools/racket/runs/379263234) that the failed tests for this recursive contract test are for jit disabled builds. I assume this could affect the asymptotic run time of the test, right? Can we use here `(eval-jit-enabled)` and only run the test if that's true?

**Comment by pmatos** [comment]
> I just noticed (https://github.com/LinkiTools/racket/runs/379263234) that the failed tests for this recursive contract test are for jit disabled builds. I assume this could affect the asymptotic run time of the test, right? Can we use here `(eval-jit-enabled)` and only run the test if that's true?

Now that's ridiculous - saying I noticed that these are failing in no-jit builds... I knew that since the beginning given I wrote it in the title of the bug. :rofl: 
The holidays were not kind to me. :smile:

**Comment by pmatos** [comment]
I added some debugging statements to the testcase and ran it in macos in GitHub Actions and in macos jitless the different between contract executions is more like 10 times slower - not maximum of 3 as the code expects.

Maybe this deserves more careful investigation, since it's something MacOS specific:
https://github.com/LinkiTools/racket/runs/399202425#step:10:49

**Comment by mflatt** [comment]
From the logs, it looks like Racket is built without optimization (i.e., without `-O2`). That will make the interpreter slow.

Maybe it makes the interpreter especially slow on Mac OS — perhaps something about relative frame sizes, causing the interpreter on Mac OS to overflow the stack a lot more. Something like that could even interfere with asymptotic complexity (if it makes GCs progressively slower to deal with progressively larger stacks).

I don't know exactly why `-O2` ends up being disabled or whether that was intentional, but using `-O2` could be something to try.

**Comment by pmatos** [comment]
> From the logs, it looks like Racket is built without optimization (i.e., without `-O2`). That will make the interpreter slow.
> 
> Maybe it makes the interpreter especially slow on Mac OS — perhaps something about relative frame sizes, causing the interpreter on Mac OS to overflow the stack a lot more. Something like that could even interfere with asymptotic complexity (if it makes GCs progressively slower to deal with progressively larger stacks).
> 
> I don't know exactly why `-O2` ends up being disabled or whether that was intentional, but using `-O2` could be something to try.

Well spotted. Compiling without optimizations is unintentional. It's due to `make CFLAGS="-Werror" -l $cpus -j $((cpus+1))` - i.e. enabling `-Werror`.

I will fix this and retry. Thanks.

**Comment by pmatos** [comment]
I have a feeling this got fixed with fb63f399ef3c543409f7e88b07ea153650d00615, so closing.
